### ISSUE TITLE
Allow for ActorRegistry.getOrRegisterActor not to create actor

### ISSUE THREAD

**Issue by victornoel** [puniverse/quasar#233]
Hi,

I would like to be able to call `ActorRegistry.getOrRegisterActor`, but in some case, I need to delegate to the passed factory the choice of creating or not the actor (basically, my use case is that I associate an actor with a object stored in a database, and if the actor does not already exists, the factory will check if the object exists in database : if not, the actor should not be created and `null`/`Optional.empty`, for example, should be returned instead).

Currently, the factory must always create an actor, couldn't we introduce another method that would accept a `Callable<Optional<Actor<Message, ?>>>` (or use the same and accept `null` return values).

**Comment by pron** [comment]
The original goal of this method is to prevent races between two threads/machines that may want to create a globally-accessed actor. For your use-case, what's wrong with calling `getActor` first, then, if `null`, test if the object is in the DB, and then, if it is, call `getOrRegisterActor` to make sure there is no race in creating the actor?

**Comment by victornoel** [comment]
It could work, but maybe the db object would be removed from the db in-between the moment I checked the existence of the actor, the existence of the object and called `getOrRegisterActor`?
I feel like this would be a problem but I didn't think about it very formally.
My intention was that only the actor tied to the database object could remove it: if the actor exists then the object is in the database, if it doesn't then either the actor wasn't created yet (and in that case I'm sure I will be the one creating it and nobody could remove the db object) or either the object is non-existent and thus querying the db will tell me that.
So I wanted to take advantage of the atomicity of `getOrRegisterActor` to ensure that there is no race for accessing/removing the db object.
Currently I simply use a lock as a singleton in my code and do what you say, but I find it inelegant (and dangerous because of the singleton, I trust your implementation of ActorRegistry more than my or my colleague's code ;) to reimplement that on top of the current implementation of `getOrRegisterActor` so I thought it could make sense for other people too to have a generalization of `getOrRegisterActor` to my use case.

**Comment by pron** [comment]
> but maybe the db object would be removed from the db in-between the moment I checked the existence of the actor, the existence of the object and called getOrRegisterActor?

That's true, except that can happen between the time your factory reads the database and returns the actor, too. 

> if the actor exists then the object is in the database

Right, but if you don't have a transaction surrounding both events -- creation of the object and of the actor -- then your algorithm is racy, unless, as you say, only the actor can remove its accompanying DB object, in which case my suggestion would work as well. In any event, it is not any more or less transactional than using `getOrRegisterActor` alone.

**Comment by pron** [comment]
P.S.

> In any event, it is not any more or less transactional than using getOrRegisterActor alone.

It may be if the factory would also *create* the DB object.

**Comment by victornoel** [comment]
The db object is never created by the factory (but most certainly the factory would be called not long after because most certainly an actor to manage it would be needed). Actually, it would make sense to also have the object created by the actor even though I could imagine use cases where the actor wouldn't be needed afterward.

> That's true, except that can happen between the time your factory reads the database and returns the actor, too.

I thought the factory was executed as part as the transactional behaviour of `getOrRegisterActor`, isn't that the case? That's maybe were my reasoning fails!

**Comment by pron** [comment]
> I thought the factory was executed as part as the transactional behaviour of getOrRegisterActor, isn't that the case?

First of all, that part is not really specified. The only thing `getOrRegisterActor` guarantees is that an actor would be returned, and no two actors would be registered. Second, even if the factory is invoked in a transaction, the transaction cannot possibly cover the database, too, as the registry knows nothing about it.

**Comment by victornoel** [comment]
Concerning the fact that the factory is not covered by the transactional behaviour of `getOrRegisterActor`, maybe then it could be interesting to include it? I don't see what would be good in creating an actor but not registering it… it would be a bit counterintuitive to users, especially in an imperative language like Java where things can happen in unforeseen places.

Concerning the transactionality of the db, I didn't meant that I could get database transactionality automagically because of the registry, but simply that I expected that if the factory is run, then it is only because there is not another actor, and thus, if there is not another actor, I know because of how I designed my actors, that it means that no object will be deleted inbetween the moment I believe the object exists and the moment I create the actor. And after that, it is this very actor that would be responsible of deleting this object, because I am sure it would be the one returned by `getOrRegisterActor`.
Concerning creation, it is not important in my use case: anyway there is no race between creation, because actor's name is the id of the object which is autoincremented in the db, so I can trust the db giving an unique id for every created object and no newly created object can replace a previously deleted object and the actor will be anyway responsible of creating the object.

I understand if you don't want to add this functionality, but I don't see how it wouldn't make my code clearer, simpler and safer :)

**Comment by pron** [comment]
> maybe then it could be interesting to include it?

That's what locks are made for, and Galaxy will have global locks. Moreover, the factory may in theory be invoked more than once. 

> it would be a bit counterintuitive to users, especially in an imperative language like Java where things can happen in unforeseen places.

Well, the same could also be said about supporting mutation in lambdas passed to Java streams. But the fact is that sometimes you want to provide certain concurrent/parallel semantics, and in imperative, non-transactional languages, that often means that users simply should be careful. Quasar precisely tries to make concurrency much friendlier to languages like Java, but that's what fibers and actors are for. I don't think it makes sense to arbitrarily designate the very specific `getOrRegisterActor` method to be a general-purpose global lock.

> Concerning the transactionality of the db,...

But your use case is precisely covered by a combined use of `getActor` and `getOrRegisterActor`. The functionality you want is already implemented. Changing `getOrRegisterActor` wouldn't support your use case -- at least as I understand it -- any better.

> I understand if you don't want to add this functionality, but I don't see how it wouldn't make my code clearer, simpler and safer

It's not that I don't want to add this particular functionality. I don't want to add *any* functionality unless it is justified. The reason it won't make your code clearer, simpler and safer is this:

* In terms of safety, the race of an object deletion between testing the DB and creating the actor is the same whether you do both in the factory or call `getActor` and then `getOrRegisterActor`.

* In terms of clarity, putting business logic in what is intended as a simple factory is less clear, IMO, then simply stating your intent in the code: test if the object exists, and if so create an actor.

* In terms of simplicity, turning a very specific method (that happens to be transactional) into a general transaction service conflates concerns and seems like the opposite of simplicity to me.

**Comment by victornoel** [comment]
Ok, it makes sense for the functional/imperative question.

For the rest I will believe you, it seems I can't understand why it's enough to combine `getActor` and `getOrRegisterActor` (without introducing my own lock, right?), but it's ok ^^

Also I get the more general reasons why it is not the best to add the functionality :)

Thanks for taking the time to explain all of this in any case, feel free to close this issue then!
